File: app/src/main/java/choruscoldchain/app/ResetPasswordActivity.java
package choruscoldchain.app;

import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.text.TextUtils;
import android.text.method.PasswordTransformationMethod;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import java.io.IOException;
import java.io.InputStream;

public class ResetPasswordActivity extends AppCompatActivity {
    private static final String TAG = "ResetPasswordActivity";
    
    private EditText etNewPassword, etConfirmPassword;
    private Button btnCreatePassword, btnBackToLogin;
    private ImageView ivChorusLogo, ivNewPasswordToggle, ivConfirmPasswordToggle;
    
    private AuthManager authManager;
    private String email;
    private String accessToken;
    private boolean isNewPasswordVisible = false;
    private boolean isConfirmPasswordVisible = false;
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_reset_password);
        
        // Get email and access token from intent
        Intent intent = getIntent();
        email = intent.getStringExtra("email");
        accessToken = intent.getStringExtra("accessToken");
        
        if (email == null || accessToken == null) {
            Toast.makeText(this, "Invalid access. Please try again.", Toast.LENGTH_LONG).show();
            finish();
            return;
        }
        
        authManager = AuthManager.getInstance(this);
        
        initViews();
        setupListeners();
        loadChorusLogo();
        setupAppId();
    }
    
    private void initViews() {
        etNewPassword = findViewById(R.id.etNewPassword);
        etConfirmPassword = findViewById(R.id.etConfirmPassword);
        btnCreatePassword = findViewById(R.id.btnCreatePassword);
        btnBackToLogin = findViewById(R.id.btnBackToLogin);
        ivChorusLogo = findViewById(R.id.ivChorusLogo);
        ivNewPasswordToggle = findViewById(R.id.ivNewPasswordToggle);
        ivConfirmPasswordToggle = findViewById(R.id.ivConfirmPasswordToggle);
    }
    
    private void loadChorusLogo() {
        // Try to load chorus logo from assets in order of preference
        String[] logoPaths = {
            "images/chorus.png",           // Main chorus logo
            "chorus_logo.png",             // Alternative chorus logo
            "images/chorusWhite.jpeg"      // White version if needed
        };
        
        for (String logoPath : logoPaths) {
            try {
                InputStream inputStream = getAssets().open(logoPath);
                Drawable drawable = Drawable.createFromStream(inputStream, null);
                if (drawable != null) {
                    ivChorusLogo.setImageDrawable(drawable);
                    inputStream.close();
                    return; // Successfully loaded, exit
                }
                inputStream.close();
            } catch (IOException e) {
                // Continue to next logo path
            }
        }
        // If no logo loaded from assets, keep the default drawable
    }
    
    private void setupListeners() {
        btnCreatePassword.setOnClickListener(v -> handleCreatePassword());
        btnBackToLogin.setOnClickListener(v -> navigateToLogin());
        
        // Password visibility toggles
        ivNewPasswordToggle.setOnClickListener(v -> toggleNewPasswordVisibility());
        ivConfirmPasswordToggle.setOnClickListener(v -> toggleConfirmPasswordVisibility());
    }
    
    private void setupAppId() {
        InstallationIdManager installationIdManager = InstallationIdManager.getInstance(this);
        TextView tvAppId = findViewById(R.id.tv_scan_app_id);
        if (tvAppId != null) {
            tvAppId.setText("App ID: " + installationIdManager.getInstallationId());
        }
    }
    
    private void toggleNewPasswordVisibility() {
        if (isNewPasswordVisible) {
            etNewPassword.setTransformationMethod(PasswordTransformationMethod.getInstance());
                            ivNewPasswordToggle.setImageResource(R.drawable.ic_visibility_off);
            isNewPasswordVisible = false;
        } else {
            etNewPassword.setTransformationMethod(null);
            ivNewPasswordToggle.setImageResource(R.drawable.ic_visibility);
            isNewPasswordVisible = true;
        }
        // Move cursor to end
        etNewPassword.setSelection(etNewPassword.getText().length());
    }
    
    private void toggleConfirmPasswordVisibility() {
        if (isConfirmPasswordVisible) {
            etConfirmPassword.setTransformationMethod(PasswordTransformationMethod.getInstance());
                            ivConfirmPasswordToggle.setImageResource(R.drawable.ic_visibility_off);
            isConfirmPasswordVisible = false;
        } else {
            etConfirmPassword.setTransformationMethod(null);
            ivConfirmPasswordToggle.setImageResource(R.drawable.ic_visibility);
            isConfirmPasswordVisible = true;
        }
        // Move cursor to end
        etConfirmPassword.setSelection(etConfirmPassword.getText().length());
    }
    
    private void handleCreatePassword() {
        String newPassword = etNewPassword.getText().toString().trim();
        String confirmPassword = etConfirmPassword.getText().toString().trim();
        
        // Validation
        if (TextUtils.isEmpty(newPassword)) {
            etNewPassword.setError("Please enter a new password");
            return;
        }
        
        if (newPassword.length() < 6) {
            etNewPassword.setError("Password must be at least 6 characters");
            return;
        }
        
        if (TextUtils.isEmpty(confirmPassword)) {
            etConfirmPassword.setError("Please confirm your password");
            return;
        }
        
        if (!newPassword.equals(confirmPassword)) {
            etConfirmPassword.setError("Passwords do not match");
            return;
        }
        
        // Clear previous errors
        etNewPassword.setError(null);
        etConfirmPassword.setError(null);
        
        // Disable button and show loading
        btnCreatePassword.setEnabled(false);
        btnCreatePassword.setText("Resetting...");
        
        // Save the access token temporarily for the reset password call
        authManager.saveAuthData(accessToken, email, null);
        
        authManager.resetPassword(newPassword, new AuthManager.AuthCallback<AuthModels.ResetPasswordResponse>() {
            @Override
            public void onSuccess(AuthModels.ResetPasswordResponse result) {
                runOnUiThread(() -> {
                    btnCreatePassword.setEnabled(true);
                    btnCreatePassword.setText("Reset Password");
                    
                    if (result.isSuccess()) {
                        Toast.makeText(ResetPasswordActivity.this, 
                            "Password reset successfully!", Toast.LENGTH_LONG).show();
                        navigateToLogin();
                    } else {
                        Toast.makeText(ResetPasswordActivity.this, 
                            result.getMessage() != null ? result.getMessage() : "Failed to reset password", 
                            Toast.LENGTH_LONG).show();
                    }
                });
            }
            
            @Override
            public void onError(String error) {
                runOnUiThread(() -> {
                    btnCreatePassword.setEnabled(true);
                    btnCreatePassword.setText("Reset Password");
                    Toast.makeText(ResetPasswordActivity.this, error, Toast.LENGTH_LONG).show();
                });
            }
        });
    }
    
    private void navigateToLogin() {
        Intent intent = new Intent(this, SignInActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(intent);
        finish();
    }
} 

File: app/src/main/java/com/minew/sensormanager/ui/activities/ResetPasswordActivity.kt
package com.minew.sensormanager.ui.activities

import android.os.Bundle
import android.text.TextUtils
import android.text.method.PasswordTransformationMethod
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.button.MaterialButton
import com.minew.sensormanager.R
import com.minew.sensormanager.auth.AuthManager
import com.minew.sensormanager.auth.AuthModels
import com.minew.sensormanager.utils.InstallationIdManager

class ResetPasswordActivity : AppCompatActivity() {
    private lateinit var etNewPassword: EditText
    private lateinit var etConfirmPassword: EditText
    private lateinit var btnCreatePassword: MaterialButton
    private lateinit var btnBackToLogin: MaterialButton
    private lateinit var ivNewPasswordToggle: ImageView
    private lateinit var ivConfirmPasswordToggle: ImageView

    private lateinit var authManager: AuthManager
    private lateinit var email: String
    private lateinit var accessToken: String
    private var isNewPasswordVisible = false
    private var isConfirmPasswordVisible = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_reset_password)

        window.decorView.systemUiVisibility =
            window.decorView.systemUiVisibility or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR

        email = intent.getStringExtra("email") ?: run {
            Toast.makeText(this, "Invalid access. Please try again.", Toast.LENGTH_LONG).show()
            finish(); return
        }
        accessToken = intent.getStringExtra("accessToken") ?: run {
            Toast.makeText(this, "Invalid access. Please try again.", Toast.LENGTH_LONG).show()
            finish(); return
        }

        authManager = AuthManager.getInstance(this)

        initViews()
        setupListeners()
        setupAppId()
    }

    private fun initViews() {
        etNewPassword = findViewById(R.id.etNewPassword)
        etConfirmPassword = findViewById(R.id.etConfirmPassword)
        btnCreatePassword = findViewById(R.id.btnCreatePassword)
        btnBackToLogin = findViewById(R.id.btnBackToLogin)
        ivNewPasswordToggle = findViewById(R.id.ivNewPasswordToggle)
        ivConfirmPasswordToggle = findViewById(R.id.ivConfirmPasswordToggle)
    }

    private fun setupListeners() {
        btnCreatePassword.setOnClickListener { handleCreatePassword() }
        btnBackToLogin.setOnClickListener { navigateToLogin() }
        ivNewPasswordToggle.setOnClickListener { toggleNewPasswordVisibility() }
        ivConfirmPasswordToggle.setOnClickListener { toggleConfirmPasswordVisibility() }
    }

    private fun setupAppId() {
        val tvAppId = findViewById<TextView>(R.id.tv_scan_app_id)
        tvAppId?.text = "App ID: ${InstallationIdManager.getInstance(this).getInstallationId()}"
    }

    private fun toggleNewPasswordVisibility() {
        if (isNewPasswordVisible) {
            etNewPassword.transformationMethod = PasswordTransformationMethod.getInstance()
            ivNewPasswordToggle.setImageResource(R.drawable.ic_visibility_off)
            isNewPasswordVisible = false
        } else {
            etNewPassword.transformationMethod = null
            ivNewPasswordToggle.setImageResource(R.drawable.ic_visibility)
            isNewPasswordVisible = true
        }
        etNewPassword.setSelection(etNewPassword.text.length)
    }

    private fun toggleConfirmPasswordVisibility() {
        if (isConfirmPasswordVisible) {
            etConfirmPassword.transformationMethod = PasswordTransformationMethod.getInstance()
            ivConfirmPasswordToggle.setImageResource(R.drawable.ic_visibility_off)
            isConfirmPasswordVisible = false
        } else {
            etConfirmPassword.transformationMethod = null
            ivConfirmPasswordToggle.setImageResource(R.drawable.ic_visibility)
            isConfirmPasswordVisible = true
        }
        etConfirmPassword.setSelection(etConfirmPassword.text.length)
    }

    private fun handleCreatePassword() {
        val newPassword = etNewPassword.text.toString().trim()
        val confirmPassword = etConfirmPassword.text.toString().trim()
        if (TextUtils.isEmpty(newPassword)) {
            etNewPassword.error = "Please enter a new password"
            return
        }
        if (newPassword.length < 6) {
            etNewPassword.error = "Password must be at least 6 characters"
            return
        }
        if (TextUtils.isEmpty(confirmPassword)) {
            etConfirmPassword.error = "Please confirm your password"
            return
        }
        if (newPassword != confirmPassword) {
            etConfirmPassword.error = "Passwords do not match"
            return
        }
        etNewPassword.error = null
        etConfirmPassword.error = null

        btnCreatePassword.isEnabled = false
        btnCreatePassword.text = "Resetting..."

        // Persist token for reset call
        authManager.saveAuthData(accessToken, email, null)
        authManager.resetPassword(newPassword, object : AuthManager.AuthCallback<AuthModels.ResetPasswordResponse> {
            override fun onSuccess(result: AuthModels.ResetPasswordResponse) {
                runOnUiThread {
                    btnCreatePassword.isEnabled = true
                    btnCreatePassword.text = "Reset Password"
                    if (result.isSuccess()) {
                        Toast.makeText(this@ResetPasswordActivity, "Password reset successfully!", Toast.LENGTH_LONG).show()
                        navigateToLogin()
                    } else {
                        Toast.makeText(this@ResetPasswordActivity, result.message ?: "Failed to reset password", Toast.LENGTH_LONG).show()
                    }
                }
            }

            override fun onError(error: String) {
                runOnUiThread {
                    btnCreatePassword.isEnabled = true
                    btnCreatePassword.text = "Reset Password"
                    Toast.makeText(this@ResetPasswordActivity, error, Toast.LENGTH_LONG).show()
                }
            }
        })
    }

    private fun navigateToLogin() {
        val intent = android.content.Intent(this, SignInActivity::class.java)
        intent.flags = android.content.Intent.FLAG_ACTIVITY_NEW_TASK or android.content.Intent.FLAG_ACTIVITY_CLEAR_TASK
        startActivity(intent)
        finish()
    }
}


